### PR TITLE
Check runcmd return status

### PR DESCRIPTION
**Pull request by mitiao**
Especially care return status for extract assets
See: https://progress.opensuse.org/issues/17902

### CODE REVIEW

File: osutils.pm [hunk 1]
```diff
@@ -74,4 +75,18 @@ sub qv($) {
...
+    bmwqemu::diag "running " . join(' ', @_);
+    my %children;
+    local $SIG{CHLD} = sub {
+        local ($!, $?);
+        while ((my $pid = waitpid(-1, WNOHANG)) > 0) {
+            delete $children{$pid};
+            cleanup_child($pid, $?);
+        }
+    };
+    return CORE::system(@_);
```
Comment (okurz) [inline osutils.pm]: Wouldn't it be safer to throw an exception if the call fails so we don't need to check the exit code explicitly on every call?

File: backend/qemu.pm [hunk 2]
```diff
@@ -271,20 +271,14 @@ sub load_snapshot {
     return $rsp;
 }
 
-sub runcmd {
-    diag "running " . join(' ', @_);
-    local $SIG{CHLD} = 'IGNORE';
-    return CORE::system(@_);
```
Comment (coolo) [inline backend/qemu.pm]: You are reverting https://github.com/os-autoinst/os-autoinst/pull/534 basically - so I object

Comment (mitiao) [reply backend/qemu.pm]: I will take `system` from CORE package back and avoid autodie

File: osutils.pm [hunk 3]
```diff
@@ -74,4 +77,20 @@ sub qv($) {
...
 
+# runcmd use SIGCHLD handler to wait all child processes at exit and reap zombies
+sub runcmd {
+    diag "running " . join(' ', @_);
+    my %children;
+    local $SIG{CHLD} = sub {
+        local ($!, $?);
+        while ((my $pid = waitpid(-1, WNOHANG)) > 0) {
+            delete $children{$pid};
+            cleanup_child($pid, $?);
```
Comment (mudler) [inline osutils.pm]: I can't find ```cleanup_child()``` in the [codebase,](https://github.com/os-autoinst/os-autoinst/search?utf8=%E2%9C%93&q=cleanup_child&type=) am i missing something?

File: isotovideo [hunk 4]
```diff
@@ -535,7 +535,11 @@ if (!$r && $completed && (my $nd = $bmwqemu::vars{NUMDISKS}) && ($bmwqemu::vars{
     }
     else {
         for my $asset (@toextract) {
-            $bmwqemu::backend->extract_assets($asset);
+            eval { $bmwqemu::backend->extract_assets($asset); };
```
Comment (mudler) [inline isotovideo]: you might want ```local ($@)``` as well

File: backend/pvm.pm [hunk 5]
```diff
@@ -45,11 +45,6 @@ sub do_start_vm {
     return {};
 }
 
-sub runcmd {
```
Comment (mudler) [inline backend/pvm.pm]: +1 for grouping common function, but afaics from the changes it's used in backends, probably best place for it should be in [baseclass](https://github.com/os-autoinst/os-autoinst/blob/master/backend/baseclass.pm)

Comment (mudler) [reply backend/pvm.pm]: Note aside: as functions in osutils are meant to be re-used, they should come up with unit tests

File: osutils.pm (hunk 3 above)
Comment (mitiao) [reply osutils.pm]: seems I can't get pid of `system` with `waitpid`, is it possible that Perl reap system subprocesses itself?
if so to use `local $SIG{CHLD} = 'DEFAULT'` should be ok here?

File: backend/pvm.pm (hunk 5 above)
Comment (mitiao) [reply backend/pvm.pm]: I moved it to baseclass

File: t/13-osutils.t [hunk 6]
```diff
@@ -130,4 +130,15 @@ subtest find_bin => sub {
 
 };
 
+subtest runcmd => sub {
+    use osutils 'runcmd';
+
+    my @cmd = ('qemu-img', 'create', '-f', 'qcow2', 'image.qcow2', '30G');
```
Comment (foursixnine) [inline t/13-osutils.t]: 30G is kind of a big number, even if it takes very short time :P can you switch it to 1G?

Comment (mitiao) [reply t/13-osutils.t]: Fine. I just thought the initial size of qcow2 format is quite small

**Comment by mitiao** [comment]
I had a test and seems it didn't display exit status of reaped childs if `system` called :( 
Also I had a few runs of failure with logs:
http://10.67.17.71/tests/109/file/autoinst-log.txt
http://10.67.17.71/tests/112/file/autoinst-log.txt

**Comment by mudler** [comment]
Yeah, because maybe what you are trying to call with ```system()``` is furthermore forking - such case you should use process groups. But that won't help if a process among them it's also declaring another process group and detaching more further.

**Comment by mitiao** [comment]
Thanks for suggestion that break my hesitation to use different approach, work to fix.

**Comment by mitiao** [comment]
runcmd failed test: http://10.67.17.71/tests/197/file/autoinst-log.txt
runcmd ok test: http://10.67.17.71/tests/196/file/autoinst-log.txt

File: osutils.pm [hunk 7]
```diff
@@ -21,13 +21,19 @@ use warnings;
 use Carp;
 use base 'Exporter';
 use Mojo::File 'path';
+use bmwqemu 'diag';
+use POSIX ':sys_wait_h';
+use Symbol 'gensym';
+use IPC::Open3;
```
Comment (okurz) [inline osutils.pm]: Do we have that package in our distris? Couldn't find it. Is it included in perl?

Comment (foursixnine) [reply osutils.pm]: @okurz Yes, it is part of the core perl distribution

Comment (mudler) [reply osutils.pm]: Take your monday morning coffee, take a breath and then, luke, use the [force](http://www.letmegooglethat.com/?q=Perl+core+module+IPC%3A%3AOpen3)

**Comment by foursixnine** [comment]
@coolo @mudler I think we can merge this already right?

**Pull request merged**